src/NRules/NRules.Json/Converters/ExpressionConverter.cs

```diff
@@ -16,72 +16,56 @@ internal class ExpressionConverter : JsonConverter<Expression>
     public override Expression Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
     {
         reader.ReadStartObject();
-        var nodeType = reader.ReadEnumProperty<ExpressionType>(nameof(Expression.NodeType), options);
-
-        switch (nodeType)
+        return reader.ReadEnumProperty<ExpressionType>(nameof(Expression.NodeType), options) switch
```

Comment:
@gritcsenko in general, I'm all in favor of using new language constructs. However, I don't think it's a good idea to convert this part to switch expression.
The main reason is that only read* side can be converted, and write* side remains a switch statement, which breaks parity between the two. Prior to your change, read and write methods were almost mirror images of each other, easy to follow and navigate.
So, can you please change it back, and fit your implementation of the new node types to the existing structure.

As a side node, as I mentioned before, mixing stylistic changes or purely syntactical code modernizations with functional changes makes it very hard for me to review, so please separate them out next time. A small code cleanup here and there is fine of course.

Reply:
@gritcsenko but let me know if you feel strongly about the switch expression, and we can discuss

Reply:
@snikolayev I'm not strong aboud switch expression. From my perspective switch expression is more compact and faster-to-type way to select a value.

As of parity, I've broke it because reading logic contained duplicate cases and double checks `ExpressionType` value. My idea was to follow DRY principle in this case. Duplication is harder to maintain and prone to errors.

Reply:
@snikolayev  I can make write to be in parity with read.
Let me try this way.

Reply:
Having read and write in parity what is left is to create a map between `ExpressionType` and specialized reader/writer to remove case duplication from DRY perspective

Reply:
Anyway, I can revert everything and just add support for `Assign` and `Block` expression types

Reply:
@gritcsenko thank you. Let me review tonight. If this addresses party concern, should be fine to leave. So let’s not rush to revert then

Reply:
@gritcsenko I'm still reviewing this change, but TBH so far, I find it much harder to read and comprehend compared to what it was. The current implementation pretty much applies the same concept across the board - either writing atomic properties/objects/arrays or delegating to a custom serializer and so on recursively. Here however, there is another level of indirection with the expression serializer, which adds complexity and also makes it different than other custom converters, e.g. RuleElementConverter.
You put quite a bit of effort into this, so I'll spend some more time trying to get a better feel for your implementation. 
It might be easier though to add block expressions support to the existing structure first, and get that merged, and tackle refactoring separately.

Reply:
@snikolayev Just in case you decide to go to simpler PR first, I've created one #321